python/pyspark/errors/__init__.py

```diff
@@ -0,0 +1,140 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from typing import Dict, Optional, Union, Any, Type
+from pyspark.errors.error_classes import ERROR_CLASSES
+
+
+class PySparkException(Exception):
+    """
+    Base Exception for handling the errors generated by PySpark
+    """
+
+    def __init__(self, error_class: str, message_parameters: Optional[Dict[str, str]] = None):
+        self._verify_error_class(error_class)
+        self._error_class = error_class
+
+        self._error_message_format = ERROR_CLASSES[error_class]
+
+        self._verify_message_parameters(message_parameters)
+        self._message_parameters = message_parameters
+
+    def _verify_error_class(self, error_class: str) -> None:
+        assert (
+            error_class in ERROR_CLASSES
+        ), f"{error_class} is not in the list of error classes: {list(ERROR_CLASSES.keys())}"
+
+    def _verify_message_parameters(
+        self, message_parameters: Optional[Dict[str, str]] = None
+    ) -> None:
+        required = set(self._error_message_format.__code__.co_varnames)
+        given = set() if message_parameters is None else set(message_parameters.keys())
+        assert given == required, f"Given message parameters: {given} , but {required} required"
+
+    def getErrorClass(self) -> str:
+        return self._error_class
+
+    def getMessageParameters(self) -> Optional[Dict[str, str]]:
+        return self._message_parameters
+
+    def getErrorMessage(self) -> str:
+        if self._message_parameters is None:
+            message = self._error_message_format()  # type: ignore[operator]
+        else:
+            message = self._error_message_format(
+                *self._message_parameters.values()
+            )  # type: ignore[operator]
+
+        return message
+
+    def __str__(self) -> str:
+        # The user-facing error message is contains error class and error message
+        # e.g. "[WRONG_NUM_COLUMNS] 'greatest' should take at least two columns"
+        return f"[{self.getErrorClass()}] {self.getErrorMessage()}"
+
+
+def notColumnOrStringError(arg_name: str, arg_type: Type[Any]) -> "PySparkException":
```

Comment:
Error names should be sufficiently descriptive of the error.

I would appreciate for any comment to improvement the naming.